pkg/proxy/topology.go

```diff
@@ -69,12 +69,38 @@ func FilterTopologyEndpoint(nodeLabels map[string]string, topologyKeys []string,
 		for _, ep := range endpoints {
 			topology := ep.GetTopology()
 			if value, found := topology[key]; found && value == topologyValue {
-				filteredEndpoint = append(filteredEndpoint, ep)
+				filteredEndpoints = append(filteredEndpoints, ep)
 			}
 		}
-		if len(filteredEndpoint) > 0 {
-			return filteredEndpoint
+		if len(filteredEndpoints) > 0 {
+			return filteredEndpoints
 		}
 	}
-	return filteredEndpoint
+	return filteredEndpoints
+}
+
+// FilterLocalEndpoint returns the node local endpoints based on configured
+// InternalTrafficPolicy.
+//
+// If InternalTrafficPolicy is Local, only return the node local endpoints.
+//
+// FilterLocalEndpoint should never be called when InternalTrafficPolicy is All.
+// If it is, all endpoints will be directly returned
+func FilterLocalEndpoint(internalTrafficPolicy *v1.ServiceInternalTrafficPolicyType, endpoints []Endpoint) []Endpoint {
+	if internalTrafficPolicy == nil || *internalTrafficPolicy == v1.ServiceInternalTrafficPolicyCluster {
```

Comment:
need to check gate?

Reply:
added logic to return the original endpoints when gate is not on.